Add NG04014 Page (#6)
ngCaribe/angular-errors: src/content/errors/04014.md

---
title: "04014"
slug: "04014"
description: NG04014 - Invalid configuration of route
author: Nelson Gutierrez
authorTwitter: nelsongutidev
date: Oct 3, 2023
---

# NG04014

`NG04014` is a runtime error that indicates there is an invalid configuration of the routes. It is aliased as `INVALID_ROUTE_CONFIG` in the `ErrorCode` enum.

### Error Information Examples

```text
Error NG04014: Invalid configuration of route 'your-route': One of the following must be provided: component, loadComponent, redirectTo, children or loadChildren.

Error NG04014: Invalid configuration of route 'your-route': redirectTo and canActivate cannot be used together. Redirects happen before activation so canActivate will never be executed
```

### Potential Causes and Solutions

#### 1. Malformed Route Configuration

**Cause**: This error may occur if a route configuration is undefined or contains syntax errors.

> **Solution**: Carefully review the route configuration for the specified path ('${your route}') and ensure it is correctly defined with the required properties. Check for missing or extra commas.

Reviewer: Should we include a "bad" code sample here?

Author: Added error examples for all of the incompatible configurations conflict scenarios

3. **Conflict Between `children` and `loadChildren`**:

   - Error: "children" and "loadChildren" cannot be used together in a route configuration.
   - Explanation: If a route has child routes defined using "children," it should not include "loadChildren" for lazy loading.

4. **Conflict Between `redirectTo` and (`component` or `loadComponent`)**:

   - Error: "redirectTo" and either "component" or "loadComponent" cannot be used together.
   - Explanation: When a route redirects to another route using "redirectTo," it should not specify a component directly using "component" or "loadComponent."

5. **Conflict Between `component` and `loadComponent`**:

   - Error: "component" and "loadComponent" cannot be used together in a route configuration.
   - Explanation: A route should either specify a component using "component" or use lazy loading with "loadComponent," but not both.

6. **Conflict Between `redirectTo` and `canActivate`**:
   - Error: "redirectTo" and "canActivate" cannot be used together.
   - Explanation: When a route redirects using "redirectTo," it should not include the "canActivate" guard because redirects occur before route activation.

> > This is particularly common for migrations from ng13 to ng14, since previously this configuration, even though not technically allowed, was permitted as it didn't throw any errors, because lazy loading and other validations were not made yet. One community work around for this case is to remove the `redirect` and instead do the redirect in the `canActivate` guard. Along with this, you will also need to add an empty `children: []` property to the route configuration.

Reviewer: ```suggestion
> This is particularly common in migrations from ng13 to ng14 because of improved validation. One community workaround for this is to remove `redirect` and do that in the `canActivate` guard instead. You will also need to add an empty `children: []` property.
```

Author: Yes, will add this instead. That's exactly what I was trying to get to, a concise yet complete explanation. thanks!
